Intrinsify Reference.reachabilityFence

iwanowww opened the pull request:
/cc hotspot-compiler

iwanowww commented:
/contributor add tholenstein

iwanowww commented:
/contributor add vlivanov

TobiHartmann commented on test/hotspot/jtreg/compiler/c2/TestReachabilityFence.java:
```diff
@@ -0,0 +1,270 @@
+            try {
+                return payload[id][offset];
+            } finally {
+//                Reference.reachabilityFence(this);
```
Drive-by comment: Is this intentionally disabled?

iwanowww replied:
Good catch. Uncommented.

(FTR I wasn't able to reproduce the problem with on-heap backing storage. Choosing between removing the test case which doesn't provoke the problem and keeping it, I chose the latter.)

chhagedorn commented on src/hotspot/share/opto/phasetype.hpp:
```diff
@@ -85,6 +85,7 @@
   flags(PHASEIDEALLOOP1,                "PhaseIdealLoop 1") \
   flags(PHASEIDEALLOOP2,                "PhaseIdealLoop 2") \
   flags(PHASEIDEALLOOP3,                "PhaseIdealLoop 3") \
+  flags(OPTIMIZE_RF,                    "Optimize Reachability Fences") \
```
Another drive-by comment: I suggest to use the full word since most people are probably not aware of this abbreviation when looking at graph dumps in IGV. You should also add this phase to the IR framework [CompilePhases](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/CompilePhase.java).

```suggestion
  flags(OPTIMIZE_REACHABILITY_FENCES,   "Optimize Reachability Fences") \
```

iwanowww replied:
IMO it doesn't clarify things much. It's enum constant name and it's not shown in IGV.

It is used in a single place [1] where it's clear what it refers to. 

[1]
```
  { // No more loop opts. It is safe to eliminate reachability fence nodes.
    TracePhase tp(_t_idealLoop);
    PhaseIdealLoop::optimize(igvn, LoopOptsEliminateRFs);
    print_method(PHASE_OPTIMIZE_RF, 2);
    if (failing())  return;
  }
```

chhagedorn replied:
That's a good point, there it really does not matter. Another thought I've just had: When you add it to the `CompilePhase` class in the IR framework and it would be nice to have the same name. There, it would be beneficial to have the full name since people then only see `CompilePhase::OPTIMIZE_RF`.

iwanowww replied:
That's fair. Renamed and added to `CompilePhase.java`.

chhagedorn replied:
Thanks for the update!

shipilev reviewed (commented):
This is an impressive piece of work. It still makes me yearn for simpler solution. (I am impressed how much it snowballed from my original `@DontInline` implementation!)

> Representing ReachabilityFence as memory barrier (e.g., MemBarCPUOrder) would solve the issue, but performance costs are prohibitively high. 

How bad is it? `MemBarCPUOrder` pinches all memory, so I assume this breaks a lot of optimizations when `RF` is sitting in the hot loop? I remember we went through a similar exercise with `Blackholes`: [JDK-8296545](https://bugs.openjdk.org/browse/JDK-8296545) -- and decided to pinch only the control. I _guessing_ this is not enough to fix RF, or is it?

Some other comments:

iwanowww commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,497 @@
+        }
+      }
+    }
+    redundant_rfs.push(rf);
```
No duplicated entries are allowed in `Compile::_reachability_fences`.

`eliminate_reachability_fences()` is called at the end of loop opts after `optimize_reachability_fences()` is done (possibly, multiple times). 

`eliminate_reachability_fences()` migrates all referents from RFs to safepoints. So, every RF node is placed on `redundant_rfs` list and there are no RF nodes left after `eliminate_reachability_fences()` is over (there's `C->reachability_fences_count() == 0` assert at the end).

After safepoints are directly linked to the referent, the corresponding RF node becomes redundant. So, I kept `redundant_rfs` as a name. I can choose a different one if you find it confusing.

Another important aspect when it comes to determining RF redundancy is that  `eliminate_reachability_fences()` takes all users into account while `optimize_reachability_fences()` trusts only `ReachabilityFence`s.

iwanowww commented on src/java.base/share/classes/java/lang/ref/Reference.java:
```diff
@@ -659,12 +659,10 @@ protected Object clone() throws CloneNotSupportedException {
      * @since 9
      */
-    @ForceInline
+    @IntrinsicCandidate
```
I'd like to use `-XX:DisableIntrinsic=_Reference_reachabilityFence` to switch to current behavior (no fence).
Also, `@DontInline` would require special handling in C1 to unconditionally inline it.

`@ForceInline` was there primarily to communicate the interaction with JVM. (Existing inlining heuristics should just unconditionally inline empty methods.) Once `@IntrinsicCandidate` is there, I don't see much value in any other annotations.

iwanowww replied:
> But the whole point of this PR is that "current behavior" is incorrect, isn't it?

Strictly speaking, current implementation has a defect and it requires a complete rewrite on C2 side to properly fix it.

Current implementation is part of JDK for a long time (since 11). It's highly unlikely it'll be backported all the way to JDK 11 and it's an open question whether it should be backported at all. So, for diagnostic purposes it makes sense to provide a way to compare old and new implementations irrespective of whether old implementation still has the bug.

> In other words, let's not rely on intrinsic to work for correctness; non-intrinsified version should be correct as well.

A question for you: do you think we should test non-intrinsified case?

Personally, I consider such requirement as way too strong. In this particular case, the method is unconditionally intrinsified in C2. If no intrinsification takes place, it's a bug. (I'm fine with adding an assert & abort compilation if C2 ever observes `Reference.reachabilityFence()` to be inlined.)

iwanowww replied:
> a) intrinsics are performance optimizations, not correctness building blocks; 

The fact that many intrinsics are performance optimizations doesn't mean all intrinsics should be.  

> I also cannot remember any existing intrinsic that is a counter-example for (a)

`Reference::get`.

> This is not about the backports, but about having the correct fallback when we need it.

I provided it as a justification why we may need access to current implementation even if we know it's not 100% correct. 

> Users then ask: "Is it safe to disable the intrinsic? Would my application crash/misbehave without it? Would it run slower?"

There's no single answer possible here and it depending on what users expect. If there's a severe regression after the fix, as the stop-the-gap solution, it's common to restore original behavior users had before the change arrived (and that's `-XX:DisableIntrinsic=_Reference_reachabilityFence`). If somebody asks for a workaround which is 100% correct, it can be suggested to also disable method inlining (`-XX:CompileCommand=dontinline,java.lang.ref.Reference::reachabilityFence`), but it won't be the mode they run before and it'll introduce additional risks for them (mostly, performance-wise).

So, there's a way to achieve both. Also, `-XX:DisableIntrinsic=_Reference_reachabilityFence` covers the most common scenario we may need and it's the simplest one from implementation POV.

iwanowww commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -3933,11 +3958,27 @@ void Compile::final_graph_reshaping_walk(Node_Stack& nstack, Node* root, Final_R
       (!UseCompressedOops && !UseCompressedClassPointers))
     return;
 
+  // Go over ReachabilityFence nodes to skip DecodeN nodes for referents.
```
`ReachabilityFence` accepts only OOPs as a referent and `DecodeNKlass` produces `Klass` pointer.

I suspect it may be the case for safepoints as well (and `is_DecodeNarrowPtr()` is a a leftover from PermGen world), but I didn't check.

shipilev replied:
Right, nevermind about `DecodeNKlass` then. My question on heap loads still stands: do we actually get `reachabilityFence(someField)` from anywhere?

iwanowww replied:
Are you asking specifically about `ReachabilityFence -> DecodeN -> LoadN` shape? Yes, it's common, especially after inlining.

eme64 replied:
@iwanowww Can you add a code comment why this is safe to look through the ReachabilityFence?

iwanowww replied:
Done.

iwanowww commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -3897,6 +3908,20 @@ void Compile::final_graph_reshaping_main_switch(Node* n, Final_Reshape_Counts& f
 // Replacing Opaque nodes with their input in final_graph_reshaping_impl(),
 // requires that the walk visits a node's inputs before visiting the node.
+
+static bool has_non_debug_uses(Node* n) {
```
I moved it to `Node::has_non_debug_uses()`.

iwanowww commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -3951,21 +3992,8 @@ void Compile::final_graph_reshaping_walk(Node_Stack& nstack, Node* root, Final_R
-        if (safe_to_skip) {
-          n->set_req(j, in->in(1));
-        }
+      if (in->is_DecodeNarrowPtr() && (is_uncommon || has_non_debug_uses(in))) {
```
Good point. Fixed.

Funnily enough, it did the right job here (but not in RF case!), because `has_non_debug_uses` reported the opposite answer to what can be expected based on its name :-)

iwanowww commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -961,8 +962,12 @@ void CallNode::extract_projections(CallProjections* projs, bool separate_io_proj
+            assert(projs->exobj == nullptr, "only one");
+            projs->exobj = e;
+          } else  {
+            // exception table for rethrow case
```
Makes sense. Added an assert.

iwanowww commented on src/hotspot/share/opto/block.cpp:
```diff
@@ -182,9 +182,11 @@ int Block::is_Empty() const {
          !get_node(end_idx)->is_Mach() &&
-         !get_node(end_idx)->is_BoxLock()) {
+         !get_node(end_idx)->is_BoxLock() &&
+         !(get_node(end_idx)->is_ReachabilityFence() && C->print_assembly())) {
```
Fair enough. Treating RFs uniformly should be fine, since they are attached to safepoints now.

iwanowww reviewed (commented):
Thanks for the feedback, Aleksey, Tobias, and Christian.

iwanowww commented:
>> Representing ReachabilityFence as memory barrier (e.g., MemBarCPUOrder) would solve the issue, but performance costs are prohibitively high.

> How bad is it? MemBarCPUOrder pinches all memory, so I assume this breaks a lot of optimizations when RF is sitting in the hot loop? I remember we went through a similar exercise with Blackholes: [JDK-8296545](https://bugs.openjdk.org/browse/JDK-8296545) -- and decided to pinch only the control. I guessing this is not enough to fix RF, or is it?

Yes, if a barrier stays inside loop body, it breaks a lot of important optimizations. It may end up almost as bad as a full-blown call (except a barrier can be moved around while a call can't). And moving a node when it depends both on control and memory is more complicated than just a CFG node. Moreover, as you can see in the proposed solution, even CFG-only representation is problematic for loop opts, so additional care is needed to ensure RFs are moved out of loops. 

As an alternative approach, I thought about reifying RF as a data node (think of `CastPP`) and then linking its referent to all safepoints it dominates after loop opts are over.  But that would only affect `optimize_reachability_fences()`. Everything else  would stay the same. So, I decided to stay with CFG-only representation for now.

shipilev commented on src/hotspot/share/opto/parse1.cpp:
```diff
@@ -1201,6 +1221,17 @@ void Parse::do_method_entry() {
   }
 
+  if (StressReachabilityFences) {
+    // Keep all oop arguments alive until method return.
```
OK!

eme64 commented on src/hotspot/share/opto/c2_globals.hpp:
```diff
@@ -73,6 +73,15 @@
   develop(bool, StressBailout, false,                                       \
           "Perform bailouts randomly at C2 failing() checks")               \
                                                                             \
+  product(bool, OptimizeReachabilityFences, true, DIAGNOSTIC,               \
+          "Optimize reachability fences")                                   \
+                                                                            \
+  product(bool, PreserveReachabilityFencesOnConstants, false, DIAGNOSTIC,   \
+          "Preserve reachability fences on constants")                      \
+                                                                            \
+  product(bool, StressReachabilityFences, false, DIAGNOSTIC,                \
+          "Randomly insert ReachabilityFence nodes")                        \
```
Drive-by sniping: what about a hello-world test where you test out these flags?

iwanowww replied:
Good idea. Added one.

eme64 commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2512,7 +2518,12 @@ void Compile::Optimize() {
   }
 
-  if (failing())  return;
+  { // No more loop opts. It is safe to eliminate reachability fence nodes.
+    TracePhase tp(_t_idealLoop);
+    PhaseIdealLoop::optimize(igvn, LoopOptsEliminateRFs);
+    print_method(PHASE_ELIMINATE_REACHABILITY_FENCES, 2);
+    if (failing())  return;
+  }
```
Might be helpful if you write in a comment if this eliminates all or just some of the reachability fences.

eme64 replied:
Can we limit it to cases where we actually have reachability fences?

iwanowww replied:
Good point. Done.

eme64 commented:
@iwanowww I reviewed half of the code. I left lots of minor comments, about code style and little code comments. They are not super important now.

I think most important to work on now is the description on the top of `reachability.cpp`. I struggle to understand the basic concepts / ideas. It would be nice if you described things in more detail, and added some definitions. I think that would be a better basis for me to jump into the implementation afterward :)

eme64 commented:
@iwanowww Let me know whenever this is ready to review again :blush:

iwanowww commented on src/hotspot/share/opto/callnode.hpp:
```diff
@@ -492,6 +492,10 @@ class SafePointNode : public MultiNode {
 
   void disconnect_from_root(PhaseIterGVN *igvn);
 
+  // Are we guaranteed that this node is a safepoint?  Not true for leaf calls and
+  // for some macro nodes whose expansion does not have a safepoint on the fast path.
+  virtual bool guaranteed_safepoint()  { return true; }
```
Well, it's a SafePointNode class after all. I lifted it from `CallNode` subclass to avoid elaborate check on SafePoint nodes (!is_Call() || as_Call() && guaranteed_safepoint()`)).

If some node extends SafePointNode, but doesn't keep JVM state, it has to communicate it to users one way or another. And changing the default doesn't improve the situation IMO: reporting a safepoint node as a non-safepoint is still a bug.

iwanowww commented on src/hotspot/share/opto/parse.hpp:
```diff
@@ -358,6 +358,7 @@ class Parse : public GraphKit {
   bool          _wrote_stable;       // Did we write a @Stable field?
   bool          _wrote_fields;       // Did we write any field?
   Node*         _alloc_with_final_or_stable; // An allocation node with final or @Stable field
+  Node*         _stress_rf_hook; // StressReachabilityFences support
```
I'd like to avoid that. `_stress_reachability_fence_hook` is way too verbose IMO. The declaration and all the accesses are accompanied by `StressReachabilityFences` which should make it clear what `rf` refers to.

iwanowww commented on test/hotspot/jtreg/compiler/c2/TestReachabilityFence.java:
```diff
@@ -0,0 +1,270 @@
+ * @bug 8290892
+ * @summary Tests to ensure that reachabilityFence() correctly keeps objects from being collected prematurely.
+ * @modules java.base/jdk.internal.misc
+ * @run main/othervm -Xbatch compiler.c2.TestReachabilityFence
```
This particular test is carefully crafted to provoke a failure when reachability fence effects aren't properly modeled. Stressing RF implementation doesn't help here.

iwanowww commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -943,11 +943,12 @@ void CallNode::extract_projections(CallProjections* projs, bool separate_io_proj
+              case CatchProjNode::fall_through_index: projs->fallthrough_catchproj = cpn; break;
+              case CatchProjNode::catch_all_index:    projs->catchall_catchproj    = cpn; break;
+              default: {
+                assert(cpn->_con > 1, ""); // exception table; rethrow case
```
Some call nodes inspected during `expand_reachability_fences` demonstrate this IR shape where some exception table projections are directly attached to the call node.

Looks like a missed case in `CallNode::extract_projections` we simply never hit before.

iwanowww commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -611,6 +611,17 @@ void CallGenerator::do_late_inline_helper() {
+
+  uint endoff = call->jvms()->endoff();
+  if (C->inlining_incrementally()) {
+    assert(endoff == call->req(), ""); // assert in SafePointNode::grow_stack
```
The assert ensures there are no reachability edges present when incremental inlining takes place. Inlining logic doesn't expect any extra edges past debug info and the comment refers to the assert which fires the first.

iwanowww commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -3933,11 +3944,27 @@ void Compile::final_graph_reshaping_walk(Node_Stack& nstack, Node* root, Final_R
+  for (int i = 0; i < C->reachability_fences_count(); i++) {
+    Node* rf = C->reachability_fence(i);
+    Node* in = rf->in(1);
+    if (in->is_DecodeN()) {
```
Ok, reshaped as you suggested.

iwanowww commented on src/hotspot/share/opto/compile.hpp:
```diff
@@ -377,6 +378,7 @@ class Compile : public Phase {
   // of Template Assertion Predicates themselves.
   GrowableArray<OpaqueTemplateAssertionPredicateNode*>  _template_assertion_predicate_opaques;
   GrowableArray<Node*>  _expensive_nodes;       // List of nodes that are expensive to compute and that we'd better not let the GVN freely common
+  GrowableArray<Node*>  _reachability_fences;   // List of reachability fences
```
Ok, done.

iwanowww commented on src/hotspot/share/opto/compile.hpp:
```diff
@@ -728,6 +732,14 @@ class Compile : public Phase {
+
+  void remove_reachability_fence(Node* n) {
+    _reachability_fences.remove_if_existing(n);
+  }
```
Done.

iwanowww commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -61,6 +61,22 @@ Node *IdealLoopTree::is_loop_exit(Node *iff) const {
+    }
+  }
+  return unique_loop_exit;
+}
```
It was easier to inspect it in the debugger. Reshaped as you suggested.

iwanowww commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1456,7 +1472,20 @@ class PhaseIdealLoop : public PhaseTransform {
+  void replace_rf(Node* old_node, Node* new_node);
+  void remove_rf(Node* rf);
+#ifdef ASSERT
+  bool has_redundant_rfs(Unique_Node_List& ignored_rfs, bool rf_only);
```
There are 2 types of methods here: internal ones (used solely in `reachability.cpp`) and those which are called from loop optimization code (`optimize_reachability_fences` and `eliminate_reachability_fences`). 

IMO it's counter-productive to repeatedly spell out what "RF" means inside `reachability.cpp`, so I kept the names intact. I split the declarations into public and private ones to stress the distinction.

iwanowww commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -977,6 +979,8 @@ void PhaseMacroExpand::process_users_of_allocation(CallNode *alloc) {
         }
         _igvn._worklist.push(ac);
+      } else if (use->is_ReachabilityFence() && OptimizeReachabilityFences) {
+        _igvn.replace_input_of(use, 1, _igvn.makecon(TypePtr::NULL_PTR)); // reset; redundant fence
```
Turned it into `ReachabilityFenceNode::clear_referent()`. Hope it makes it clearer.

iwanowww commented on src/hotspot/share/opto/node.hpp:
```diff
@@ -693,8 +697,9 @@ class Node {
-        DEFINE_CLASS_ID(Initialize,       MemBar, 0)
-        DEFINE_CLASS_ID(MemBarStoreStore, MemBar, 1)
+        DEFINE_CLASS_ID(Initialize,        MemBar, 0)
+        DEFINE_CLASS_ID(MemBarStoreStore,  MemBar, 1)
```
Fixed.

iwanowww commented:
@eme64 please, take another look. Thanks!

eme64 reviewed (commented):
Thanks for all the updates. I went over all your responses quickly, but still need to read through the description in `reachability.cpp` now.

I'll do another pass over everything once you address my responses ;)

dean-long commented:
@iwanowww , do you have a test that shows constant oops are a problem?  My initial impression is that PreserveReachabilityFencesOnConstants shouldn't be needed, because any oops referenced during the compile should go into the ciEnv metadata[] and then into the nmethod oops.  So GC can't reclaim these oops because the nmethod keeps references to them.

dean-long commented on src/hotspot/share/opto/callGenerator.cpp:
```diff
@@ -611,6 +611,20 @@ void CallGenerator::do_late_inline_helper() {
   Compile* C = Compile::current();
+
+  uint endoff = call->jvms()->endoff();
+  if (C->inlining_incrementally()) {
+    // No reachability edges should be present when incremental inlining takes place.
+    // Inlining logic doesn't expect any extra edges past debug info and fails with
+    // an assert in SafePointNode::grow_stack.
+    assert(endoff == call->req(), "reachability edges not supported");
```
We also use this area past endoff() for storing the "ex_oop" (see for example GraphKit::has_saved_ex_oop()).  Are ex_oop and reachability edges mutually exclusive?

iwanowww replied:
Yes, ex_oop and reachability edges are mutually exclusive, but there's no conflict. ex_oop is kept during parsing while reachability edges stay attached to RF nodes until loop optimizations are over (and no inlining can happen anymore).

iwanowww commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -5302,6 +5329,16 @@ void PhaseIdealLoop::build_and_optimize() {
     C->set_major_progress();
   }
+
+  // Loop opts are over. It is safe to get rid of all reachability fence nodes
+  // and migrate reachability edges to safepoints.
+  if (OptimizeReachabilityFences && !C->major_progress() && C->reachability_fences_count() > 0 && _mode == LoopOptsDefaultFinal) {
+    Compile::TracePhase tp2(_t_reachability);
+    eliminate_reachability_fences();
+    if (C->failing())  return;
+    C->print_method(PHASE_ELIMINATE_REACHABILITY_FENCES, 2);
+    assert(C->reachability_fences_count() == 0, "no RF nodes allowed");
+  }
```
> Why are you checking for _mode == LoopOptsDefaultFinal and not for LoopOptsEliminateRFs?

The intention is to avoid an extra `PhaseIdealLoop` construction pass solely for `LoopOptsEliminateRFs` purposes when there's an empty pass during normal flow of loop optimizations.   

`LoopOptsEliminateRFs` is performed as the last resort when there was no previous pass to piggyback on.

iwanowww replied:
Maybe `LoopOptsEliminateRFs` should stress that it is intended to happen as the very last step in the flow of loop optimizations. Or, something happening after all other loop optimizations are over. I'll think more about it.

From code perspective, what makes things more complicated is that  `PhaseIdealLoop` instance is hidden in `PhaseIdealLoop::optimize()`, so shaping it as a step in loop opts pipeline feels like the most appropriate thing to do.

eme64 replied:
@iwanowww It is the last step in the pipeline, but the pipeline could get executed again, right? So then you may think that you have reached the last step in the pipeline, but then in the next pipeline execution, you might have already eliminated the RF, and now you would do some loop-opts that you should not. That's what I'm worried about. Can we have some assert for that?

iwanowww commented:
@dean-long 

> because any oops referenced during the compile should go into the ciEnv metadata[] and then into the nmethod oop

That's not how it behaves in practice. The OOPs observed during compilation don't necessarily end up in nmethod metadata unless there're explicit usages. 

> do you have a test that shows constant oops are a problem?

I do. Just pushed one example as `test/hotspot/jtreg/compiler/c2/TestReachabilityFenceOnConstant.java`.

iwanowww commented:
> Extra compile-time due to extra loop-opts round. https://github.com/openjdk/jdk/pull/25315#discussion_r2330176841 . It used to be a 20% increase, now you managed to make it only 10%. Still considerable. 

FTR 10% increase in loop opts time is observed with `-XX:+StressReachabiltyFences`.

> All of it just to call get_ctrl(referent) in enumerate_interfering_sfpts.

Well, I wouldn't frame it in such a way. RF elimination transformation relies on dominance information computed by `PhaseIdealLoop` to produce control input for each referent. And there are many other transformations under `PhaseIdealLoop` which "just" rely on dominance info it produces.

eme64 commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -966,6 +969,8 @@ void PhaseMacroExpand::process_users_of_allocation(CallNode *alloc) {
         }
         _igvn._worklist.push(ac);
+      } else if (use->is_ReachabilityFence() && OptimizeReachabilityFences) {
+        use->as_ReachabilityFence()->clear_referent(_igvn); // redundant fence
```
Ah ok. As long as it only iterates over the result cast, that is good :)

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,541 @@
+ * Hence, we eliminate RF nodes and transfer their referents to corresponding safepoints (phase #2).
+ * When safepoints are pruned, corresponding reachability edges also go away.
+ *
+ * (c) Unfortunately, it's not straightforward to stay with safepoint-attached representation till the very end,
+ * because information about derived oops is attached to safepoints in a similar way. So, for now RFs are
+ * rematerialized at safepoints before RA (phase #3).
```
Maybe we should do that effort soon, otherwise we just keep heaping up tech dept :/

iwanowww commented:
/label remove core-libs graal hotspot
/label add hotspot-compiler

dean-long commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -1223,23 +1223,29 @@ bool ConnectionGraph::reduce_phi_on_safepoints_helper(Node* ophi, Node* cast, No
     // See more details of these fields in the declaration of SafePointScalarMergeNode
     sfpt->add_req(nsr_merge_pointer);
     sfpt->add_req(selector);
+    sfpt->jvms()->set_endoff(sfpt->req());
```
This seems like a subtle change that deserves a comment.  Is it changing the behavior?  Now these two edges are considered part of scalar/debug edges, when before they weren't?

iwanowww replied:
> Is it changing the behavior?

It doesn't. The sole purpose of the change is to please the assert in `create_scalarized_object_description()`.

At the end of successful call, `create_scalarized_object_description()` extends debug info to cover all newly added inputs. And the failure is non-recoverable. So, the very first iteration of the loop extends debug info beyond those 2 edges.  

I'll add a comment.

iwanowww commented:
@eme64 @dean-long I addressed your feedback in the latest version. Please, take another look. Thanks!

(It passed testing up to hs-tier10 w/ `-XX:+StressReachabilityFences`.)   

In particular, I reverted the optimization to piggy-back on last PhaseIdealLoop pass to eliminate RF nodes. Filed JDK-8370127 to address it separately.

I filed the following bugs/enhancements for follow-up work:
1. JDK-8370127 "C2: Improve ReachabilityFence elimination"
2. JDK-8370129 "C2: Support ReachabilityFence with constant referent"
3. JDK-8370131 "C2: Improve ReachabilityFence test coverage"
4. JDK-8370132 "C2: Enforce post-loop opts phase"
5. JDK-8370133 "C2: Manage non-debug safepoint edges in structural manner"
6. JDK-8370137 "C2: Ensure dependent memory accesses can't float past ReachabilityFence"

iwanowww commented:
Any reviews, please?

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,554 @@
+// RF is redundant for some referent oop when the referent has another user which keeps it alive across the RF.
+// In terms of dominance relation it can be formulated as "a referent has a user which is dominated by the redundant RF".
+// Until loop opts are over, only RF nodes are considered as usages (controlled by rf_only flag).
+static bool is_redundant_rf_helper(Node* ctrl, Node* referent, PhaseIdealLoop* phase, PhaseGVN& gvn, bool rf_only) {
```
Nit: `_helper` is fine if it is used as some internal method, i.e. only `is_redundant_rf` uses `is_redundant_rf_helper`. But it seems you are using it from different places. Can you find a better name?

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,554 @@
+    if (m != nullptr && m->is_boxing_method()) {
+      return true;
+    }
+  }
```
Can you explain in a code comment?

eme64 reviewed (commented):
That is how far I got today. Only was able to look at details, I'll have to get the overview tomorrow or next week.

Thanks for keeping on working on this, this is a complex one!

I left off at `reachability.cpp` at `PhaseIdealLoop::insert_rf`, marking it for myself ;)

iwanowww commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -61,6 +61,24 @@ Node *IdealLoopTree::is_loop_exit(Node *iff) const {
+    } else if (head()->is_OuterStripMinedLoop()) {
+      return head()->as_OuterStripMinedLoop()->outer_loop_exit();
+    } else {
+      // For now, conservatively report multiple loop exits exist.
```
The outer check is `is_loop()`, but special cases are for counted loops and strip-mined outer loops. So, any non-counted loop handling should land here.

iwanowww commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,554 @@
+    } else if (sfpt->is_CallStaticJava() && sfpt->as_CallStaticJava()->is_uncommon_trap()) {
+      return false; // uncommon traps are exit points
+    }
+    return true;
```
`is_significant_sfpt()` encodes a white list consisting of cases which can be safely ignored when it comes to reachability tracking. An overlooked case is a missed optimization opportunity. 

> But can you be more specific?

Are you suggesting to expand the comment or change the name?

Speaking of the name, it's a local definition inside `reachabiltiy.cpp`:
```
// Detect safepoint nodes which are important for reachability tracking purposes.
static bool is_significant_sfpt(Node* n) {
```

`Significant` term is declarative. Alternatively, an imperative term (like, "ignore" and `ignore_sfpt`) can be used. But I find it the current version clearer.

eme64 replied:
> is_significant_sfpt() encodes a white list consisting of cases which can be safely ignored when it comes to reachability tracking. An overlooked case is a missed optimization opportunity.

Sounds good. Can you add a code comment for that, please?

Ok, I'm fine with keeping the name. But it might make sense to link to where the "significance" term is defined. Because otherwise it is a concept without any clear definition, and hard for the reader to understand. You have to infer the definition from the usage, and that often leads to unclear definitions that shift over time, and eventually the concept even is incoherent. A clear definition can also help if we have a bug: we can clearly see what we missed and what we might have to do to fix it.

eme64 replied:
Nice! `is_interfering_sfpt_candidate` is definitively a big improvement :)

iwanowww commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1140,6 +1146,17 @@ class PhaseIdealLoop : public PhaseTransform {
   }
 
+  void remove_dead_node(Node* dead) {
+    assert(dead->outcnt() == 0 && !dead->is_top(), "node must be dead");
```
> We should probably also not call this on a CFG node, otherwise we might destroy the "ctrl forwarding"

Interesting. RF is a CFG node. How does broken ctrl forwarding manifest? Does `VerifyLoopOptimizations` catch it? I didn't see any failures with `-XX:+VerifyLoopOptimizations`.

eme64 replied:
> Does VerifyLoopOptimizations catch it?

Probably not, because it is not very strong yet. I will probably soon work on it again, to make sure we have stronger invariants, or at least enforcing our implicit invariants ;)

In your case, you probably are using your new `remove_dead_node` in the way it "should" be used. But since this is a public API, someone will probably come along in the future and use it in unintended ways. That's why I'm asking for more asserts that at least block some wrong usages ;)

eme64 replied:
Nice, thanks for the update!

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -1223,23 +1223,32 @@ bool ConnectionGraph::reduce_phi_on_safepoints_helper(Node* ophi, Node* cast, No
-    uint merge_idx      = (sfpt->req() - jvms->scloff());
-    int debug_start     = jvms->debug_start();
+
+    sfpt->remove_non_debug_edges(non_debug_edges_worklist);
```
@iwanowww Thanks for the comments!

I'm still not 100% happy with it. It really feels like we are introducing some tech-debt here. What should the next person do who also needs to attach something else to the SafePoint? It is also easy to miss places where we have to special case the extra edges. I don't have a solution here, I'm just not extremely satisfied. Is there a better long-term solution?

iwanowww replied:
I don't think we can improve things much without paying existing tech debt related to managing non-debug edges. 
The two cases I spotted relates to scalarization and is not aware about non-debug inputs a safepoint may have. Both implementations assume there's only debug info attached and tweak it by appending new edges and extending debug info range to cover them. Without a proper API to manage safepoint-attached debug information, it'll always be fragile. 

JDK-8370133 "C2: Manage non-debug safepoint edges in structural manner" should address the root cause and introduce a proper way to work with safepoint-attached debug and non-debug information.

iwanowww commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -5109,8 +5115,12 @@ void PhaseIdealLoop::build_and_optimize() {
-    assert(do_expensive_nodes, "why are we here?");
-    if (process_expensive_nodes()) {
+    assert(do_expensive_nodes || do_optimize_reachability_fences, "why are we here?");
+    if (do_optimize_reachability_fences && optimize_reachability_fences()) {
```
Done.

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+    if (m != nullptr && m->is_boxing_method()) {
+      return true; // ignore fences on boxed primitives produced by valueOf methods
+    }
+  }
```
Nice, thanks for adding the comment. I'm trying to understand the reason why that is ok.

So someone would have set a RF for a boxed primitive. But we don't expect anything to ever be attached to a boxed primitive, and so we can just ignore these RF? Is that the reason? Might be worth writing it in a code comment explicitly.

iwanowww replied:
I stumbled upon it while inspecting IR with `-XX:+StressReachabiltiyFences` when RF is added for auto-boxed argument.  (`is_boxing_method()` matches `valueOf` factory, not an explicitly allocated box).

Thinking more about it, there's a way to observe the absence of RF on a auto-boxed instance when (1) value is out of range for internal caches; and (2) there's java.lang.ref.Reference instance registered for it.

So, I moved the logic under `StressReachabilityFences` instead.

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+
+// Linearly traverse CFG upwards starting at n until first merge point.
+// All encountered safepoints are recorded in safepoints list.
+static void linear_traversal(Node* n, Node_Stack& worklist, VectorSet& visited, Node_List& safepoints) {
```
The second comment line does not sound accurate, we don't collect ALL, only the candidates. Maybe we can find a better method name, and even remove that comment because of it?

Given the more useful sub query `is_interfering_sfpt_candidate`, I think we could name this method something like `collect_interfering_sfpt_candidates`. Or is it very important that this is a linear traversal vs some other traversal we could choose from?

eme64 replied:
Hmm, but this here is only a component of the `enumerate_interfering_sfpts` below, which has essencially that name.

So maybe it should be `collect_interfering_sfpt_candidates_for_node` here and just `collect_interfering_sfpt_candidates` below?

iwanowww replied:
I updated the comment and did some renaming.

`enumerate_interfering_sfpts_linear_traversal` computes the set of interfering safepoints for some RF node, not just candidates. It's only `is_interfering_sfpt_candidate()` which does some filtering ignoring the context (the RF node a safepoint can interfere with).

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+  VectorSet visited;
+  visited.set(referent_ctrl->_idx); // end point
+
+  Node_Stack stack(0);
```
`ResouceMark`?

iwanowww replied:
It's hard to place a nested ResourceMark because there are dynamically reallocated data structures with different life cycles. Instead, I moved temporary data structure allocations up in the call chain and made them shared across all RF nodes.

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+    }
+    if (!is_redundant_rf(rf, false /*rf_only*/)) {
+      Node_List safepoints;
+      enumerate_interfering_sfpts(rf, this, safepoints);
```
Could this explode if we have a lot of RF in the graph? Do we need a ResouceMark, or reuse the `safeponts` node list?

Imagine something like this:
```
referent
if (flag1) { something } else { something }
...
if (flag100) { something } else { something }
if (x1) { RF(referent); }
...
if (x100) { RF(referent); }
```
So we would call `enumerate_interfering_sfpts` 100x, and then traverse the graph with about 100-400 nodes each time. You can see how that grows quadratically. Maybe that's fine for runtime, but is it also ok for memory?

eme64 replied:
And what if we find a lot of SafePoints for each RF? Do we end up attaching quadratically many referent edges over all?

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+  Compile::TracePhase tp(_t_reachability_eliminate);
+
+  assert(OptimizeReachabilityFences, "required");
+  assert(C->post_loop_opts_phase(), "required");
```
You could give the reason in the assert message ;)

iwanowww replied:
Those are essentially preconditions. I'm not a fan of duplicating information over and over again. Assert messages won't add much beyond what the condition already says unless the message explains why precondition is there. But that would explode message size.

I can use `precond` here if you are against generic messages in asserts.

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+        }
+      }
+    }
+    redundant_rfs.push(rf);
```
Another question:
What if `!is_redundant_rf` and we found no sfpt candidates? Is it ok to just eliminate the rf here?

eme64 replied:
I suppose yes, because no GC could have happened since, right?

iwanowww commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+// Linearly traverse CFG upwards starting at n until first merge point.
+// All encountered safepoints are recorded in safepoints list.
+static void linear_traversal(Node* n, Node_Stack& worklist, VectorSet& visited, Node_List& safepoints) {
+  for (Node* ctrl = n; ctrl != nullptr; ctrl = ctrl->in(0)) {
```
`linear_traversal` in `enumerate_interfering_sfpts` is an optimization for CFG traversal. Only Region nodes end up on worklist while the rest can be traversed in a linear fashion. 

> Why not just add all CFG nodes on the worklist, would that not simplify the graph a little?
Then on each iteration you have to check for Region node and handle it specially. The perceived complexity just pops out in a different place. I started with a generic graph traversal version and then found it clearer to separate Region handling and avoid worklist for all the other CFG nodes.

iwanowww commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+
+  ResourceMark rm;
+  Unique_Node_List redundant_rfs;
+  Node_List worklist;
```
Sounds good, I applied your suggestion.

> Also: why do we need this worklist, and not just attach the referent to the sfpt eagerly?

There's an assert to ensure there are no non-debug edges on discovered safepoints. It'd be harder to ensure the invariant if safepoints are modified during the analysis.  

> And: can it be that we add the same pair multiple times? Is that intentional?

Good catch. I added the missing check guarding `add_req` call.

iwanowww commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+
+// RF is redundant for some referent oop when the referent has another user which keeps it alive across the RF.
+// In terms of dominance relation it can be formulated as "a referent has a user which is dominated by the redundant RF".
+// Until loop opts are over, only RF nodes are considered as usages (controlled by rf_only flag).
```
> Can you say why until after loop opts are over only RF are considered?

It is intended to ensure a RF is superseded by another RF. We can't guarantee that non-RF users won't go away any time before loop opts are over, so there won't be anything keeping the referent alive. 

But now I don't see why a similar problem doesn't affect RFs as well. While RFs don't go away on their own, they can be pruned as part of dead code elimination.

I'll rethink how RF redundancy is defined.

> How does this play with allocation elimination etc? 

This method works on RF nodes and, hence, it is applicable until RF nodes are eliminated after loop opts are over. 

RF nodes are ignored during scalarization attempts, so are reachability edges. When an allocation is scalarized, either all RFs having it as a referent or all of it's reachability edges are removed.

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,555 @@
+  VectorSet visited;
+  visited.set(referent_ctrl->_idx); // end point
+
+  Node_Stack stack(0);
```
We already have something for node uses with `Node::visit_uses`. You could have a similar one for visiting inputs, and then limit to `CFG` nodes.

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -323,90 +211,85 @@ bool PhaseIdealLoop::optimize_reachability_fences() {
-        worklist.push(referent);
-        worklist.push(loop_exit);
-        redundant_rfs.push(rf);
+    assert(!rf->is_redundant(igvn()), "required");
```
Why can we assume this? Is this guaranteed by IGVN?

iwanowww replied:
Yes, the assumption is that last IGVN pass removed all redundant RFs and the graph hasn't changed for new cases to occur.

eme64 commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -418,72 +301,109 @@ static void enumerate_interfering_sfpts(ReachabilityFenceNode* rf, PhaseIdealLoo
+
 // Phase 2: migrate reachability info to safepoints.
 // All RFs are replaced with edges from corresponding referents to interfering safepoints.
 // Interfering safepoints are safepoint nodes which are reachable from the RF to its referent through CFG.
```
Seems you don't do it for ALL any more, you drop those that `dominates_another_rf`. You should probably adapt the comment here.

eme64 commented:
You have a few tests already, but I'd love to see some IR tests. You could even check for the presence of `ReachabilityFenceNode` during some phase and then see if it goes away. Nice would be if we could even track if a SafePoint has a RF edge attached, but not sure how easy that is.

It would allow us not only to check for correctness, and hoping that we would catch incorrect cases with a crash/wrong result. But it would allow us to verify the graph, including the optimizations.

What do you think?

iwanowww commented on src/hotspot/share/opto/reachability.cpp:
Ok, adjusted.

iwanowww commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,476 @@
+  if (lpt->_reachability_fences == nullptr) {
+    lpt->_reachability_fences = new Node_List();
+  }
+  lpt->_reachability_fences->push(new_rf);
```
Done.

iwanowww commented:
Reviews, please?

vnkozlov commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -4021,21 +4055,8 @@ void Compile::final_graph_reshaping_walk(Node_Stack& nstack, Node* root, Final_R
-          for (uint i = 0; i < in->outcnt(); i++) {
-            Node* u = in->raw_out(i);
-            if (!u->is_SafePoint() ||
-                (u->is_Call() && u->as_Call()->has_non_debug_use(n))) {
```
There was bug here! should be `in` instead of `n`.  Your change for this part is correct.

vnkozlov reviewed (commented):
In general changes seem fine.

There are changes which seems not related to reachability fence code. May be test and push them separately.

Also update copyright year.

iwanowww commented:
@dean-long 

> For solution 1, slides mention "negatively affects loop opts", which I don't see described in the comments.

"Attempt №1" referred to a solution where ReachabilityFence mimics Blackhole node.
Since it is part of CFG, it it hinders some transformations happening during loop optimizations (e.g., auto-vectorization requires no control flow in loop body). Also, it keeps the referent alive irrespective of whether there are any interfering safepoints or not. So, refined approach introduced:
(1) optimization for RFs with loop-invariant referents during loop opts;
(2) switch to safepoint-attached representation once loop opts are over.   

> For redundant RFs, if the kept RF is going to extend the live range of the reference, then the redundant RF with shorter live range seems harmless.

Indeed, redundant RFs are harmless. But reducing the total number of RFs speeds up optimization/transformation passes over them. So, I decided to eliminate them anyway.

dean-long commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -2508,12 +2515,21 @@ void Compile::Optimize() {
+    if (failing())  return;
+    assert(_reachability_fences.length() == 0 || PreserveReachabilityFencesOnConstants, "no RF nodes allowed");
+  }
+
```
It's a little confusing that "expand reachability fences" is used for both "edges --> RFs" and "RFs --> edges".

iwanowww commented on src/hotspot/share/opto/compile.cpp:
```diff
@@ -3968,11 +3984,29 @@ void Compile::final_graph_reshaping_walk(Node_Stack& nstack, Node* root, Final_R
 
+  // Go over ReachabilityFence nodes to skip DecodeN nodes for referents.
+  // The sole purpose of RF node is to keep the referent oop alive and
+  // decoding the oop for that is not needed.
```
The optimization is not specific to RF. Originally, it covered safepoints, but the same reasoning apply to RFs as well. 

IMO final graph reshaping phase is an appropriate place for it. Theoretically, it can be done earlier (after parsing is over), but then no new DecodeN usages should be introduces. Otherwise, the transformation has to be reversed.

dean-long commented on src/hotspot/share/opto/loopnode.hpp:
```diff
@@ -1167,6 +1175,18 @@ class PhaseIdealLoop : public PhaseTransform {
+    Node* c = get_ctrl(dead);
+    IdealLoopTree* lpt = get_loop(c);
+    _loop_or_ctrl.map(dead->_idx, nullptr); // This node is useless
+    if (!lpt->is_root()) {
```
Is this an optimization?  The node can be on this list, so why not remove it?

iwanowww replied:
I was under the impression that it's not the case, but I double-checked and indeed the node can be there. Fixed (here and in another place).

dean-long commented on src/hotspot/share/opto/node.hpp:
```diff
@@ -706,6 +710,7 @@ class Node {
       DEFINE_CLASS_ID(MemBar,      Multi, 3)
         DEFINE_CLASS_ID(Initialize,       MemBar, 0)
         DEFINE_CLASS_ID(MemBarStoreStore, MemBar, 1)
+      DEFINE_CLASS_ID(ReachabilityFence, Multi, 4)
```
ReachabilityFenceNode is not a subclass of MultiNode.

dean-long commented on src/hotspot/share/opto/callnode.cpp:
```suggestion
          CatchProjNode *cpn = nullptr;
```

dean-long commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,507 @@
+          // valid_length_test_input is appended during macro expansion at the very end
+          int last_idx = sfpt->req() - 1;
+          extra_edge = sfpt->in(last_idx);
+          sfpt->del_req(last_idx);
```
```suggestion
          rf_offset++;
```

dean-long commented on src/hotspot/share/opto/loopTransform.cpp:
```diff
@@ -50,17 +50,64 @@
+        continue; // skip
+      } else if (ctrl->is_Proj()) {
+        if (ctrl->is_IfProj() || ctrl->Opcode() == Op_SCMemProj || ctrl->Opcode() == Op_Proj) {
+          // skip simple control projections
```
To match style.
```suggestion
          continue; // skip simple control projections
```

iwanowww commented on src/hotspot/share/opto/loopnode.cpp:
```diff
@@ -7467,7 +7508,7 @@ void LoopTreeIterator::next() {
   if (_curnt->_child != nullptr) {
     _curnt = _curnt->_child;
-  } else if (_curnt->_next != nullptr) {
+  } else if (_curnt != _root && _curnt->_next != nullptr) {
```
I vaguely remember there was a bug there, but since I don't use `LoopTreeIterator` anymore, it's unrelated to the change. I'll file a separate bug for it.

dean-long commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,507 @@
+
+// Start offset for reachability info on a safepoint node.
+static uint rf_base_offset(SafePointNode* sfpt) {
+  return sfpt->jvms()->oopoff();
```
Of the three, debug_end() seems the best name to be used for "non debug start", plus it has extra asserts.

dean-long commented on src/hotspot/share/opto/callnode.cpp:
```diff
@@ -964,11 +964,10 @@ void CallNode::extract_projections(CallProjections* projs, bool separate_io_proj
-              projs->catchall_catchproj = cpn;
+            switch (cpn->_con) {
+              case CatchProjNode::fall_through_index: projs->fallthrough_catchproj = cpn; break;
+              case CatchProjNode::catch_all_index:    projs->catchall_catchproj    = cpn; break;
```
Thinking about this a little more, if given the choice between an exception handler with _con == 1 and handler_bci >= 0 vs a rethrow handler with _con > 1 and handler_bci -1, the latter feels like the better choice.  However, are we guaranteed that either path won't be found unreachable/dead in the future?  Because RF is using the result not to rewire the call, but to add a node that needs to stay around for correctness.

dean-long commented on src/hotspot/share/opto/reachability.cpp:
```diff
@@ -0,0 +1,507 @@
+static Node* sfpt_ctrl_out(SafePointNode* sfpt) {
+  if (sfpt->is_Call()) {
+    CallProjections callprojs;
+    sfpt->as_Call()->extract_projections(&callprojs, false /*separate_io_proj*/, false /*do_asserts*/);
```
Yes, it's the existing call sites I'm worried about.  Instead of returning a boolean and adding assert in callers, how about a default param `bool allow_incomplete = true` (or maybe `allow_handlers`), and sfpt_ctrl_out can set it to false.

iwanowww commented:
/template append

iwanowww commented:
Thanks for the reviews!

/integrate
